### PR TITLE
optimize: refactor InsertExecutor

### LINKED ISSUE #2042: refactor: InsertExecutor refactor
## Why you need it?
 Is your feature request related to a problem? Please describe in details

InsertExecutor refactor

## How it could be?
A clear and concise description of what you want to happen. You can explain more about input of the feature, and output of it.

### PR DESCRIPTION
**Pull request by jsbxyyx**
### Ⅰ. Describe what this PR did
Refactor insert executor.

### Ⅱ. Does this pull request fix one issue?

fixes #2042

### CODE REVIEW

**Comment by zjinlei** [comment]
please restore UpdateExecutorTest.java, SelectForUpdateExecutorTest.java, PlainExecutorTest.java and DeleteExecutorTest.java.

**Comment by jsbxyyx** [comment]
@zjinlei fix done.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        if (pkMetaMap.size() != 1) {
+            throw new NotSupportYetException();
+        }
+        ResultSet genKeys = statementProxy.getTargetStatement().getGeneratedKeys();
+        List<Object> pkValues = new ArrayList<>();
+        while (genKeys.next()) {
+            Object v = genKeys.getObject(1);
+            pkValues.add(v);
+        }
+        return pkValues;
```
Comment (jaspercloud) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java:65]: 测试结果取不到值没有next

Comment (jaspercloud) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java:65]: insert into access_log (id) values (nextval('access_log_id_seq'))

**Comment by jsbxyyx** [comment]
base on #2301

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java (hunk 1 above)
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java:66]: fix it.

File: sqlparser/seata-sqlparser-core/src/main/java/io/seata/sqlparser/struct/SqlDefaultExpr.java [hunk 2]
```diff
@@ -25,7 +26,7 @@
     /**
      * Get SqlDefaultExpr.
      *
-     * @return the SqlDefaultExpr
+     * @return the sqlDefaultExpr
```
Comment (zjinlei) [inline sqlparser/seata-sqlparser-core/src/main/java/io/seata/sqlparser/struct/SqlDefaultExpr.java:29]: ?

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java [hunk 3]
```diff
@@ -84,7 +85,8 @@
                 SQLRecognizer sqlRecognizer = sqlRecognizers.get(0);
                 switch (sqlRecognizer.getSQLType()) {
                     case INSERT:
-                        executor = new InsertExecutor<>(statementProxy, statementCallback, sqlRecognizer);
+                        executor = InsertExecutorFactory.createInsertExecutor(statementProxy, statementCallback,
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java:88]: can use EnhancedServiceLoader.load directly.

Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/ExecuteTemplate.java]: done.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseInsertExecutor.java [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+        return containsPK(insertColumns);
+    }
+
+    /**
+     * judge sql specify column
+     * @return true: contains column. false: not contains column.
+     */
+    protected boolean containsColumns() {
+        SQLInsertRecognizer recognizer = (SQLInsertRecognizer) sqlRecognizer;
+        List<String> insertColumns = recognizer.getInsertColumns();
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseInsertExecutor.java:94]: Just judge that it is not empty, can simplify the implementation.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/mysql/MySQLInsertExecutor.java [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author jsbxyyx
+ */
+@LoadLevel(name = JdbcConstants.MYSQL, scope = Scope.PROTOTYPE)
+public class MySQLInsertExecutor extends BaseInsertExecutor implements Defaultable {
```
[the remark this answers is not in the document]
Comment (jsbxyyx) [reply rm-datasource/src/main/java/io/seata/rm/datasource/exec/mysql/MySQLInsertExecutor.java:43]: mysql has default keyword too.
image

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseInsertExecutor.java [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+                                idx = totalPlaceholderNum - currentRowPlaceholderNum + pkIndex;
+                            }
+                            ArrayList<Object> parameter = parameters[idx];
+                            pkValues.addAll(parameter);
+                        } else {
+                            pkValues.add(pkValue);
+                        }
+                    }
+                }
+            }
```
Comment (zjinlei) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/BaseInsertExecutor.java:182]: Reduce coupling and implement in 3 steps
1.getInsertRows
2.get pkIndex list from rows
3.get pkValue list from parameters

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/oracle/OracleInsertExecutor.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+            pkValues = getGeneratedKeys();
+        } else if (pkValues.size() == 1 && pkValues.get(0) instanceof Null) {
+            throw new NotSupportYetException("oracle not support null");
+        }
+        return pkValues;
+    }
+
+    @Override
+    public String getSequenceSql(SqlSequenceExpr expr) {
+        return "SELECT " + expr.getSequence() + ".currval FROM DUAL";
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/oracle/OracleInsertExecutor.java:77]: How about turn the sql into constant.

File: rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        String function = org.apache.commons.lang.StringUtils.substringBetween(columnDef, "", "(");
+        if (StringUtils.isBlank(seq)) {
+            throw new ShouldNeverHappenException("get primary key value failed, cause columnDef is " + columnDef);
+        }
+        return getPkValuesBySequence(new SqlSequenceExpr("'" + seq + "'", function));
+    }
+
+    @Override
+    public String getSequenceSql(SqlSequenceExpr expr) {
+        return "SELECT currval(" + expr.getSequence() + ")";
```
Comment (l81893521) [inline rm-datasource/src/main/java/io/seata/rm/datasource/exec/postgresql/PostgresqlInsertExecutor.java:101]: How about turn the sql into constant.

**Pull request merged**